Helm chart for OpenTelemetry Collector

dmitryax opened the pull request:
**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/227

More TODO:
- [x] Other outputs
- [x] Docs
- [ ] Tests
- [ ] Fill values.schema.json

dmitryax committed:
Helm chart for OpenTelemetry Collector

naseemkullah commented on helm-charts/opentelemetry-collector/values.yaml:
```diff
@@ -0,0 +1,111 @@
+    hostPort: 9411
+    protocol: TCP
+
+# Configuration for agent OpenTelemetry Collector daemonset, enabled by default
+agentCollector:
+  enabled: false
+
+  resources:
+    limits:
+      cpu: 256m
+      memory: 512Mi
+
+  # Configuration override that will be merged into the agent collector default config
+  configOverride: {}
+
+# Configuration for standalone OpenTelemetry Collector deployment, disabled by default
+standaloneCollector:
+  enabled: true
```
The way the chart values are laid out make it seem like if one wanted both collector types in their cluster they could achieve this via 1 helm release, which may very well be the intention in which case you it may be desirable to embed more options in addition to `resources` such as independent `serviceAccount`s, `securityContext`s and so on. However if the intention is that one helm release result in either a standalone deployment OR an agent daemonset being deployed then I would move `resources` to top level, and use a top level option such as `collectorType: agent|standalone` or `kind: daemonset/deployment`.  WDYT?

dmitryax replied:
The intention is to be able to deploy both at the same time. It's a pretty common use case to have traces flowing through agents and concentrated/batched in a standalone collector. The k8s example we have shows exactly that https://github.com/open-telemetry/opentelemetry-collector/blob/master/examples/k8s.yaml. Not sure if both need to be enabled by default. I'd rather have just deamonset enabled by default. 

Having it in one release make it easier to configure traces/metrics flowing through the standalone collector by default it that one is enabled. That is achieved using this config override https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1026/files#diff-88a82221bc3c6b6f6ee0d5d04617e12e. And it involves less manual steps. So I'd prefer using one helm release over two.

I agree that it has to be flexible enough so every part (agent or standalone collector) gets it's own independent serviceAccounts, securityContexts etc. I'll define them once here at top level, but will make it possible to override for each independently.

Please let me know what do you think

CC @open-telemetry/collector-approvers

naseemkullah replied:
ok sgtm good to know one release is meant to deploy both standalone and agent services

dmitryax replied:
I've just realized that I already made all the options like serviceAccount etc overridable for each of the component using this trick:  https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1026/files#diff-03f976ab7d04d60820f32aa59b25577eR24-R26 I needed agent/standalone specific options along with top level options in the pod template. It might be not very obvious, I'll put more descriptions in the values

naseemkullah replied:
👏

naseemkullah commented on helm-charts/opentelemetry-collector/templates/hpa.yaml:
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.autoscaling.enabled }}
```
This condition should be combined with that of the standalone collector being true or we can wind up with a daemonset and an hpa.

dmitryax replied:
thanks. addressed

naseemkullah reviewed (changes requested):
Looks really good!

Left some comments about wether or not if both collector types are desired, how this should be achieved either by one helm release or two. Also that autoscaling enablement should also depend on standalone being deployed.

tigrannajaryan commented:
@dmitryax should we close this since there is a separate repo now?

dmitryax commented:
@tigrannajaryan I'd rather finalize the initial PR here to be approved, then close and resubmit it in the helm repo as is.

naseemkullah commented on helm-charts/opentelemetry-collector/values.yaml:
```diff
@@ -0,0 +1,111 @@
+  # Configuration override that will be merged into the agent collector default config
+  configOverride: {}
+
+# Configuration for standalone OpenTelemetry Collector deployment, disabled by default
```
Comment seems wrong about disabled by default

dmitryax replied:
thanks. comment is right. defaults were wrong fixed

naseemkullah commented on helm-charts/opentelemetry-collector/values.yaml:
```diff
@@ -0,0 +1,111 @@
+    hostPort: 9411
+    protocol: TCP
+
+# Configuration for agent OpenTelemetry Collector daemonset, enabled by default
```
Comment seems wrong about enabled by default

dmitryax replied:
same here

dmitryax committed:
address PR comments

dmitryax committed:
fix typos

dmitryax committed:
Fix checksum annotation for agent

dmitryax committed:
Remove resource detection processor to use core image by default

dmitryax committed:
Reorganize global output to exporter by telemetry

dmitryax committed:
Fix usage of configOverride

dmitryax committed:
Clarify that all the top level options can be overridden by component

dmitryax committed:
Update appVersion

dmitryax committed:
Disable host metrics by default, simplify config building

naseemkullah reviewed (approved):
lgtm, I look forward to trying this out once the helm repo is set up and hosting it!
I do have one non-blocking nit: WDYT about renaming the `agent-collector` component/`agentCollector` value to just `agent`? 
The current name makes it sounds like it collects agents.

dmitryax commented:
> I do have one non-blocking nit: WDYT about renaming the agent-collector component/agentCollector value to just agent?
> The current name makes it sounds like it collects agents.

I agree it sounds a bit confusing. I was trying to have some consistency so `standaloneCollector`/`agentCollector` . Do you suggest `standaloneCollector`/`agent` instead?

naseemkullah commented:
> > I do have one non-blocking nit: WDYT about renaming the agent-collector component/agentCollector value to just agent?
> > The current name makes it sounds like it collects agents.
> 
> I agree it sounds a bit confusing. I was trying to have some consistency, so used this pair `standaloneCollector`/`agentCollector` . Do you suggest `standaloneCollector`/`agent` instead?

Ah I see, I would be prefer with any of these:

- `standaloneCollector`/`agent`
- `standalone`/`agent`
-`collector`/`agent`


The latter is probably my favorite, despite the service itself having collector in the name but any of these are nicer imho

dmitryax commented:
We might want to introduce another name for the standalone collector which doesn't conflict with otel collector itself. So far I've heard versions like `gateway` or `concentrator` ...

@open-telemetry/collector-approvers any thoughts/suggestions on the naming for collector deployed as daemonset against collector deployed as standalone deployment? ^

owais commented:
Some SDK libraries also use the term agent to refer to the auto-instrumentation agent for example, Java: https://github.com/open-telemetry/opentelemetry-java-instrumentation/

I guess .Net might do the same once they have auto-instrumentation. So probably better to not call the collector an agent? I don't think it's a huge deal but would be nice if we could use a specific name that can't be confused with other things in the opentelemetry ecosystem.

keitwb commented:
I like what @dmitryax did with `standaloneCollector/agentCollector`, in that it has the term `collector` in both.  However, `standalone` isn't very descriptive since all of the collectors are technically "stand alone" (not paired with anything else to do it's job).  

One thing that has always been confusing to me is the totally different terms used for the same application (_agent_ vs _collector_) to describe different configurations.  Probably calling both of them some variation of _collector_ is the clearest.  Maybe `centralCollector` and `hostCollector`.

naseemkullah commented:
imho, if we go down the route of having collector in both components' names, `collectorAgent` rolls off the tongue a little nicer than `agentCollector` which sounds like something that collects agents, that's if we even continue to use the word `agent` though

dmitryax commented:
BTW choosing the naming we also need to consider how to differentiate resource names like configmaps etc for each of the components. I added `-agent` suffix for `agentCollector` components, and `standaloneCollector` goes with no suffix.

dmitryax committed:
Fix config overrides precedence

dmitryax committed:
Remove specific config for hostmetrics but provide a way to enable it

Keeping specific config for hostmetrics doesn't allow to disable particular scrapers

dmitryax committed:
Handle better disabling specific ports, more comments

dmitryax committed:
Add README.md

dmitryax commented:
I'm closing this PR in favor of https://github.com/open-telemetry/opentelemetry-helm-charts/pull/2 .
I kept naming as is for now, and moved our discussion to the issue https://github.com/open-telemetry/opentelemetry-helm-charts/issues/3
